File: examples/python-project/src/repository/user_repository.py
import sqlite3  # OK: Repository can access database
import logging
from typing import Optional, Dict


class UserRepository:
    def __init__(self):
        # OK: Repository can have database connection
        self.conn = sqlite3.connect('users.db')
        self.cursor = self.conn.cursor()
        self._create_table()
    
    def _create_table(self):
        # OK: Repository can execute SQL
        self.cursor.execute('''
            CREATE TABLE IF NOT EXISTS users (
                id TEXT PRIMARY KEY,
                name TEXT NOT NULL
            )
        ''')
        self.conn.commit()
    
    def get_by_id(self, user_id: str) -> Optional[Dict]:
        # OK: Repository can execute SQL queries
        self.cursor.execute("SELECT id, name FROM users WHERE id = ?", (user_id,))
        
        # This is fine - using logger
        logging.info(f"Fetching user from database: {user_id}")
        
        row = self.cursor.fetchone()
        if row:
            return {"id": row[0], "name": row[1]}
        return None
    
    def create(self, user_id: str, name: str) -> bool:
        # OK: Repository can execute SQL statements
        try:
            self.cursor.execute(
                "INSERT INTO users (id, name) VALUES (?, ?)",
                (user_id, name)
            )
            self.conn.commit()
            return True
        except sqlite3.IntegrityError:
            logging.error(f"User {user_id} already exists")
            return False

File: examples/go-project/pkg/repository/user_repository.go
package repository

import (
	"database/sql" // OK: Repository can access database
	"log"
)

type UserRepository struct {
	db *sql.DB // OK: Repository can have database reference
}

func NewUserRepository(db *sql.DB) *UserRepository {
	return &UserRepository{db: db}
}

func (r *UserRepository) GetByID(id string) (*User, error) {
	// OK: Repository can execute SQL queries
	row := r.db.QueryRow("SELECT id, name FROM users WHERE id = ?", id)

	// This is fine - using logger
	log.Printf("Fetching user from database: %s", id)

	var user User
	err := row.Scan(&user.ID, &user.Name)
	if err != nil {
		return nil, err
	}

	return &user, nil
}

func (r *UserRepository) Create(user *User) error {
	// OK: Repository can execute SQL statements
	_, err := r.db.Exec("INSERT INTO users (id, name) VALUES (?, ?)", user.ID, user.Name)
	return err
}

type User struct {
	ID   string
	Name string
}
